File: ruby/scrabble-score/scrabble_score.rb
class Scrabble
  LETTER_SCORES = [
    [%w{A E I O U L N R S T}, 1],
    [%w{D G}, 2],
    [%w{B C M P}, 3],
    [%w{F H V W Y}, 4],
    [%w{K}, 5],
    [%w{J X}, 8],
    [%w{Q Z}, 10]
  ]

  LETTER_SCORE_MAP = Hash[
    LETTER_SCORES.map { |letters, score|
      letters.map { |letter| [letter, score] }
    }.flatten(1)
  ]

  attr_reader :word

  def initialize(word)
    @word = word
  end

  def score
    return 0 if word.nil?
    word
      .upcase
      .scan(/[A-Z]/)
      .map { |letter| LETTER_SCORE_MAP[letter] }
      .reduce(0, :+)
  end

  def self.score(word)
    new(word).score
  end
end

File: ecmascript/scrabble-score/scrabble-score.js
const SCORES = {
  1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
  2: ["D", "G"],
  3: ["B", "C", "M", "P"],
  4: ["F", "H", "V", "W", "Y"],
  5: ["K"],
  8: ["J", "X"],
  10: ["Q", "Z"]
};

const LETTER_SCORES = Object.keys(SCORES)
  .map(n => SCORES[n].map(c => [c, n]))
  .reduce((acc, x) => acc.concat(x), [])
  .reduce((acc, [c, n]) => { acc[c] = Number(n); return acc; }, {});

const score = (word) => {
  return [...(word || '').toUpperCase()].reduce((acc, x) => acc + LETTER_SCORES[x], 0);
};

export default score;
